### PR TITLE
Add optional data-array-parameter to vectors, matrices, quaternion and color constructors

### PR DESCRIPTION
**Pull request by aidinabedi**
This PR, implements a popular and well requested feature that allows developers to optionally pass a single pure data array as only parameter to all relevant math (and color) primitives (`Vec2`/`3`/`4`, `Mat2`/`3`, `Quat` and `Color`). The passed data array is copied in the constructor (as optimized as possible) and thus used to populate all components instead of manually having to write:
```javascript
var data = [1, 2, 3, 4];
var vector = new pc.Vec4(data[0], data[1], data[2], data[3]); // even worse for Mat3 and Mat4 :)
```
This PR finally lets us just write:
```javascript
var data = [1, 2, 3, 4];
var vector = new pc.Vec4(data); // works for all vectors, matrices, quaternion and color!
```

In addition to alleviating a big annoyance :), this PR also removes the usage of `arguments` in the `Vec2` and `Vec4` constructor, to minimize chance of JS function deoptimisations. And finally as a bonus, the PR also fully documents all original parameters in each constructor (which where missing), in addition to the new optional array parameter of course.

Summary:
- Completely backwards-compatible.
- Full API documentation.
- Tested.
- New optional APIs:
- - `new pc.Vec2(data)`
- - `new pc.Vec3(data)`
- - `new pc.Vec4(data)`
- - `new pc.Mat3(data)`
- - `new pc.Mat3(data)`
- - `new pc.Quat(data)`
- - `new pc.Color(data)`

### CODE REVIEW

File: src/core/color.js [hunk 1]
```diff
@@ -18,10 +18,17 @@ pc.extend(pc, (function () {
         this.data = new Float32Array(this.buffer, 0, 4);
         this.data3 = new Float32Array(this.buffer, 0, 3);
 
-        this.data[0] = r || 0;
-        this.data[1] = g || 0;
-        this.data[2] = b || 0;
-        this.data[3] = a !== undefined ? a : 1.0;
+        if (r && (r.length === 3 || r.length === 4)) {
```
Comment (vkalpias) [inline src/core/color.js]: Better cache `r.length` in a variable instead of calling it twice?

Comment (aidinabedi) [inline src/core/color.js]: Done! See commit https://github.com/playcanvas/engine/pull/846/commits/ca7c04b3032becb3e39632e219a6572cafe2063c

**Comment by Maksims** [comment]
I have tried few existing projects with this PR, and it seems like it has some issues. Not sure what exactly is the problem, but running any project from Editor, leads to black screen.

To test things from Editor, serve engine from localhost like so:
```
cd build/output
python -m SimpleHTTPServer 51000
```
And then open any Editor scene link or Launch, with HTTP (not SSL), and add argument: `?use_local_engine=true`.

This will use local engine and allows you to test your engine changes on real existing projects.

**Comment by aidinabedi** [comment]
Hello @Maksims!

Thanks for letting me know about the issue! I'm sorry, we must have missed it when upgrading the old feature branch to the latest stable. I apologize for the oversight. I believe we've fixed it now. Could you please try it again?

**Pull request merged**